### PR TITLE
DAOS-3822 control: Add version subcommand to utilities

### PR DESCRIPTION
**Pull request by mjmac**
Adds a "version" subcommand to daos_server, daos_agent, and dmg.
Also adds version information to daos_agent, but it will only be
displayed if invoked directly.

The version displayed is supplied by the build system.

### CODE REVIEW

**Comment by mjmac** [comment]
Sorry about the rebase/force push -- I had added a commit to address the build failure which turned out to be a duplicate of brian's fix, and I couldn't cleanly remove my commit from the patch stack.

**Comment by mjmac** [comment]
> 
> 
> Is there a reason that there looks to be a lot of refactoring of unrelated code in this? Also maybe drop in the comment what you'd expect to see for the version. If it will be the tag for main branches or if it will be a git hash or what not.

Not sure what you mean by unrelated code. Examples?

**Comment by mjmac** [comment]
Example outputs:
```
[mjmac@localhost daos]$ daos_server version
daos_server version 0.7.0
[mjmac@localhost daos]$ daos_admin
ERROR: daos_admin (version 0.7.0) may only be invoked by daos_server
[mjmac@localhost daos]$ daos_server version
daos_server version 0.7.0
[mjmac@localhost daos]$ dmg version
dmg version 0.7.0
[mjmac@localhost daos]$ daos_agent version
Starting daos_agent:
daos_agent version 0.7.0
[mjmac@localhost daos]$ daos_admin
ERROR: daos_admin (version 0.7.0) may only be invoked by daos_server
```

File: src/control/cmd/agent/main.go [hunk 1]
```diff
@@ -40,19 +41,30 @@ import (
...
-	Insecure   bool   `short:"i" long:"insecure" description:"have agent attempt to connect without certificates"`
-	RuntimeDir string `short:"s" long:"runtime_dir" description:"Path to agent communications socket"`
-	LogFile    string `short:"l" long:"logfile" description:"Full path and filename for daos agent log file"`
+	AllowProxy bool       `long:"allow-proxy" description:"Allow proxy configuration via environment"`
+	Debug      bool       `short:"d" long:"debug" description:"Enable debug output"`
+	JSON       bool       `short:"j" long:"json" description:"Enable JSON output"`
+	ConfigPath string     `short:"o" long:"config-path" description:"Path to agent configuration file" default:"etc/daos_agent.yml"`
+	Insecure   bool       `short:"i" long:"insecure" description:"have agent attempt to connect without certificates"`
+	RuntimeDir string     `short:"s" long:"runtime_dir" description:"Path to agent communications socket"`
+	LogFile    string     `short:"l" long:"logfile" description:"Full path and filename for daos agent log file"`
```
Comment (bfaccini) [inline src/control/cmd/agent/main.go]: I agree with David, it is strange that these lines appear as modified

File: src/control/SConscript [hunk 2]
```diff
@@ -68,13 +68,13 @@ def install_go_bin(denv, gosrc, gopath, libs, name, install_name):
...
                  cmd_build(path, gosrc,
-                           "go install %s" % install_src))
+                           'go install -ldflags "-X main.daosVersion=%s" %s' % (DAOS_VERSION, install_src)))
     denv.InstallAs(install_bin, bin_path)
 #pylint: enable=too-many-arguments
 
 def scons():
     """Execute build"""
-    Import('env', 'prereqs')
+    Import('env', 'prereqs', 'DAOS_VERSION')
```
Comment (bfaccini) [inline src/control/SConscript]: Is this required ? I did not have to add this in src/utils/SConscript for daos tool, as it is defined and passed from main SConstruct

Comment (mjmac) [reply src/control/SConscript]: If I don't add that, then I get this error when building:
```
NameError: global name 'DAOS_VERSION' is not defined:
  File "/home/mjmac/go/src/github.com/daos-stack/daos/SConstruct", line 310:
    scons()
...
etc.
```

File: src/control/cmd/agent/main.go (hunk 1 above)
Comment (mjmac) [reply src/control/cmd/agent/main.go]: Adding the version subcommand to this struct caused a reformat due to the width.

Comment (bfaccini) [reply src/control/cmd/agent/main.go]: Oops, I have missed the formatting change... Thx and sorry Mike!

File: src/control/SConscript (hunk 2 above)
Comment (bfaccini) [reply src/control/SConscript]: Ah ok, may be it is because you refer to it in line #71 ...

**Comment by mjmac** [comment]
> 
> 
> as @dpquigl was alluding to, the fact you are refactoring the use of appSetup, which is unrelated to adding a new subcommand, should at least be mentioned in the commit message

... But it's not unrelated. As was discussed in the Go channel, there wasn't any way to add the version subcommand without the refactoring, because otherwise the user would have to supply a hostlist in order to execute the version command.

File: src/control/cmd/agent/main.go (hunk 1 above)
Comment (bfaccini) [reply src/control/cmd/agent/main.go]: Oops, sorry to have missed that!

File: src/control/SConscript (hunk 2 above)
Comment (bfaccini) [reply src/control/SConscript]: Ah ok, could be due to the fact you reference it in line 71 ...

**Comment by dpquigl** [comment]
@mjmac I was referring to the same thing Tom was talking about. It wasn't obvious from the changes that you needed to do the appSetup refactor. If that was discussed in the go channel please try to reflect that in the commit message so its a bit more obvious.

**Comment by johannlombardi** [comment]
Seems to be taking forever. Is it still in the queue?

**Comment by mjmac** [comment]
It's third in line now.

**Comment by mjmac** [comment]
I hesitate to recommend force-landing after the previous day's shenanigans, but I will say that this has passed functional testing, and the changes shouldn't have any effect on hardware testing. It's so close now, though, so the safest thing would be to wait.

**Pull request merged**